### ISSUE TITLE
When to unfold logical connectives?

### ISSUE THREAD

**Issue by mtzguido** [FStarLang/FStar#1482]
None of these are trivially normalized-away. The first one also includes a `hasEq int` which is expected, but the rest should be normalized away. (Maybe the `hasEq int` too if we add some support for that.)
```
module Bug1482

#set-options "--no_smt"
let _ = assert_norm (1 = 1)
let _ = assert_norm (1 == 1)
let _ = assert_norm (1 === 1)
```

```
examples/bug-reports/bug1482.fst(4,0-4,27): (Error 298) Q = (*Subtyping check failed; expected type Prims.eqtype; got type Type0*) Prims.hasEq Prims.int /\
(forall (any_result: Type0).
    Prims.int == any_result ==>
    (forall (any_result: Prims.bool).
        true == any_result ==>
        (forall (any_result: Prims.logical).
            Prims.l_True == any_result ==> x: Prims.unit{Prims.equals true true})))
A query could not be solved internally, and --no_smt was given
examples/bug-reports/bug1482.fst(5,0-5,28): (Error 298) Q = Prims.auto_squash (x: Prims.unit{Prims.equals 1 1})
A query could not be solved internally, and --no_smt was given
examples/bug-reports/bug1482.fst(6,0-6,29): (Error 298) Q = Prims.auto_squash (x: Prims.unit{Prims.h_equals 1 1})
A query could not be solved internally, and --no_smt was given
Verified module: Bug1482 (8 milliseconds)
3 errors were reported (see above)
```

**Comment by mtzguido** [comment]
Dumping some thoughts.

This happens since the equality connectives (`==`, and `===`) get unfolded to their definitions as squashes of inductive types, and we don't have rules to simplify them. In general, we don't want that unfolding to occur, since it messes up not only this but the SMT encoding and intelligibility of the terms as well. In Meta-F*, we explicitly disable them (cf. the `tac_opaque`) attribute. Why not always? We could opt-in to unfolding them when needed.

**Comment by mtzguido** [comment]
This was recently fixed by 3a21716e14736fe04d368c00567f703e814c4ce5
